packages/product/src/services/product-module-service.ts

```diff
@@ -405,6 +398,229 @@ export default class ProductModuleService<
+          const variantsToCreate: ProductTypes.CreateProductVariantDTO[] = []
+          const variantsToUpdate: ProductTypes.UpdateProductVariantDTO[] = []
+          const existingVariants = existingProductVariantsMap.get(productId)
+
+          variants.forEach((variant) => {
+            const isVariantIdDefined = ("id" in variant) && isDefined(variant.id)
+            const existingProductVariant = existingVariants
+              ?.find((existingVariant) => isVariantIdDefined && (existingVariant.id === variant.id))
+
+            if (isVariantIdDefined && !existingProductVariant) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `ProductVariant "${variant.id}" not found for product "${productId}"`
+              )
+            }
+
+            if (isVariantIdDefined) {
+              variantsToUpdate.push(variant as ProductTypes.UpdateProductVariantDTO)
+            } else {
+              variantsToCreate.push(variant as ProductTypes.CreateProductVariantDTO)
+            }
+
+            const variantOptions = variant.options?.map((option, index) => {
+              const productOption = productOptions[index]
+              return {
+                option: productOption,
+                value: option.value,
+              }
+            })
+
+            if (variantOptions) {
+              variant.options = variantOptions
+            }
+          })
+
+          productVariantsToCreateMap.set(productId, variantsToCreate)
+          productVariantsToUpdateMap.set(productId, variantsToUpdate)
+
+          const variantsToUpdateIds = variantsToUpdate.map(v => v?.id) as string[]
+          const existingVariantIds = existingVariants?.map(v => v.id) || []
+          const variantsToUpdateSet = new Set(variantsToUpdateIds)
+
+          productVariantIdsToDelete = productVariantIdsToDelete.concat([
+            ...new Set(
+              existingVariantIds.filter(x => !variantsToUpdateSet.has(x))
+            )
+          ])
+        }
+
+        await Promise.all(
+          [...productVariantsToCreateMap].map(async ([productId, variants]) => {
+            return await this.productVariantService_.create(
+              productByIdMap.get(productId)!,
+              variants as unknown as ProductTypes.CreateProductVariantOnlyDTO[],
+              sharedContext
+            )
+          })
+        )
+
+        await Promise.all(
+          [...productVariantsToUpdateMap].map(async ([productId, variants]) => {
+            return await this.productVariantService_.update(
+              productByIdMap.get(productId)!,
+              variants as unknown as ProductVariantServiceTypes.UpdateProductVariantDTO[],
+              sharedContext
+            )
+          })
+        )
+
+        if (productVariantIdsToDelete.length) {
+          await this.productVariantService_.delete(productVariantIdsToDelete, sharedContext)
+        }
```

Comment:
todo: those 3 actions does not overlap, I suggest to make them concurrently such as 

```ts
await Promise.all([
  [...productVariantsToCreateMap].map(async ([productId, variants]) => {
    return await this.productVariantService_.create(
      productByIdMap.get(productId)!,
      variants as unknown as ProductTypes.CreateProductVariantOnlyDTO[],
      sharedContext
    )
  }),
[...productVariantsToUpdateMap].map(async ([productId, variants]) => {
            return await this.productVariantService_.update(
              productByIdMap.get(productId)!,
              variants as unknown as ProductVariantServiceTypes.UpdateProductVariantDTO[],
              sharedContext
            )
          }),
await this.productVariantService_.delete(productVariantIdsToDelete, sharedContext)
].flat())
```